raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,568 @@
+    /// to provide the last balance proof, which modifies the closing participant's state.
+    /// Can be called multiple times, by anyone.
+    function updateTransfer(
+        uint channel_identifier,
+        uint64 nonce,
+        uint256 transferred_amount,
+        bytes32 locksroot,
+        bytes32 additional_hash,
+        bytes signature)
+        isClosed(channel_identifier)
+        stillTimeout(channel_identifier)
+        public
+    {
+        address partner_address;
+
+        partner_address = recoverAddressFromSignature(
+            channel_identifier,
+            nonce,
+            transferred_amount,
+            locksroot,
+            additional_hash,
+            signature
+        );
+
+        uint64 old_nonce = channels[channel_identifier].participants[partner_address].nonce;
+        bytes32 old_locksroot = channels[channel_identifier].participants[partner_address].locksroot;
+
+        // This will reset the transferred amount, invalidating any unlocked locks that were
+        // unlocked but not included in the new locksroot
+        updateParticipantStruct(
+            channel_identifier,
+            partner_address,
+            nonce,
+            locksroot,
+            transferred_amount
+        );
+
+        // Clean up storage for the old (nonce, locksroot)
+        // after the locksroot has been replaced.
+        // This cannot be used anymore to unlock locks anyway.
+        bytes32 old_key = keccak256(old_nonce, old_locksroot);
+        delete channels[channel_identifier].participants[partner_address].withdrawn_locks[old_key];
+
+        TransferUpdated(channel_identifier, msg.sender);
+    }
+
+    /// @notice Unlocks a pending transfer and increases the partner's transferred amount
+    /// with the transfer value. A lock can be unlocked only once per participant.
+    // Anyone can call unlock a transfer on behalf of a channel participant.
+    function unlock(
+        uint channel_identifier,
+        address partner,
+        uint64 expiration_block,
+        uint amount,
+        bytes32 hashlock,
+        bytes merkle_proof,
+        bytes32 secret)
+        stillTimeout(channel_identifier)
+        public
+    {
+        bytes32 key;
+        bytes32 computed_locksroot;
+        bytes memory locked_encoded;
+
+        // Check that the partner is a channel participant.
+        require(channels[channel_identifier].participants[partner].initialized);
+
+        Participant storage partner_state = channels[channel_identifier].participants[partner];
+
+        // An empty locksroot means there are no pending locks
+        require(partner_state.locksroot != 0);
+
+        // The lock must not have expired, it does not matter how far in the
+        // future it would have expired
+        require(expiration_block >= block.number);
+        require(hashlock == keccak256(secret));
+
+        locked_encoded = encodeLock(expiration_block, amount, hashlock);
```

Comment:
`keccak256(expiration_block, amount, hashlock);` is sufficient here.

Reply:
In decodeLock is says the lock has 72 bytes -> it concatenates, not hashes to bytes32 
```
function decodeLock(bytes lock)
        pure
        internal
        returns (uint64 expiration, uint256 amount, bytes32 hashlock)
{
       require(lock.length == 72);
       ....
}
```

Reply:
I don't understand your point.

Reply:
So, we currently have:
```
(expiration, amount, hashlock) = decodeLock(locked_encoded);
...
h = computeMerkleRoot(locked_encoded, merkle_proof);
```
I now need to create `locked_encoded` from those 3 values. You said a `keccak256` was sufficient there, but a `keccak256` would just be a hash of the 3 values => `bytes32`. And I see that I need to concatenate the values to get 72 bytes

Reply:
The merkle tree is composed of the locks hashes:
> I now need to create locked_encoded from those 3 values.

`locked_encoded` is a just the three values concatenated `expiration_block || amount || hashlock`. After it is concatenated together we take the hash of it:

https://github.com/raiden-network/raiden/blob/f32a14713aafea62a6f5b7616c8799ed8f8f6a32/raiden/smart_contracts/NettingChannelLibrary.sol#L439

IOW, the merkle tree is composed of the hash of the encoded locks, and that is equivalente to `keccak256(expiration_block, amount, hashlock);`

> keccak256(...) returns (bytes32): compute the Ethereum-SHA-3 (Keccak-256) hash of the (tightly packed) arguments

https://solidity.readthedocs.io/en/develop/miscellaneous.html#global-variables

Reply:
Argh, did not see the `h = keccak256(lock);` in the `computeMerkleRoot`. Ok, got it.

Reply:
fixed https://github.com/raiden-network/raiden/pull/1286/commits/7e8cb11449b5b88709c54761f4b274764fe9d8e5